### PR TITLE
[Discussion] AFL without* globals

### PR DESCRIPTION
**Pull request by domenukk**
This is a first draft how a non-global AFL could look like.
It now passes all test cases but still has many rough edges.

The new structs are still stored globally for some destructuring magic, sigalarm is not handled for more than one instance of afl/the forksever and I may have missed some global vars.

Please try it out, test the speed, see how you like the naming, etc. :)

### CODE REVIEW

File: unicorn_mode/samples/persistent/harness.c [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
    Run under AFL as follows:
 
-   $ cd <afl_path>/unicorn_mode/samples/persistent/
+   $ cd <afl_path>/afl->unicorn_mode/samples/persistent/
```
Comment (vanhauser-thc) [inline unicorn_mode/samples/persistent/harness.c]: really  "afl->unicorn_mode"?

Comment (domenukk) [reply unicorn_mode/samples/persistent/harness.c]: Hehe that one was a renaming bug. 🙃
It's in both harness files, too.

**Comment by vanhauser-thc** [comment]
in general looks good to me.

Some comments:
 - could also already define MAP_SIZE as a variable where it is used? for future usage where its variable. looks like it fits well within this project item
 - We should not merge it the next two weeks :) and first do proper testing that everything works. basically when test.sh runs through fine on all our target platforms incl. *BSD
 - I am also curious if there is a performance penalty. is afl-fuz compiling and working so far?

**Comment by domenukk** [comment]
A variable `MAP_SIZE` is a good idea and is already partly in, namely the following are in `sharemem_t`:
` size_t         size_alloc; /* actual allocated size */
  size_t         size_used;  /* in use by shmem app */
`
But are not yet used widely.

afl-fuzz works. I'd be cool to leverage googles https://google.github.io/fuzzbench/ to benchmark it.

**Comment by domenukk** [comment]
Turns out getting rid of MAP_SIZE is a bigger undertaking: 

- There is `MAP_SIZE_POW2` also
- Some arrays are allocated based on `MAP_SIZE`  
- Probably other things

I think the better way is to keep MAP_SIZE almost everywhere and instead use `shm->size_used` (and have `size_used_pow2`) where applicable.

**Comment by domenukk** [comment]
Welp... Something seems to be wrong, still :/

image

**Comment by domenukk** [comment]
Nevermind, that was only in my local changes. not pushing :)

**Comment by domenukk** [comment]
The latest commit added linked lists and tidied the apis a bit, if anybody now wants to build a proper timer system, feel free (@gsoc students?)

**Comment by vanhauser-thc** [comment]
I did some speed tests, it is about a 0.3-0.7% speed decrease which is not much in my opinion. But of course the new alarm handling can also impact it further when it is there.

**Comment by domenukk** [comment]
> I did some speed tests, it is about a 0.3-0.7% speed decrease which is not much in my opinion. But of course the new alarm handling can also impact it further when it is there.

A new alarm handling won't make it into this PR, I think we're better off starting to use threads instead of signals after all.
Most timeouts waiting for can be replaced with `select`.

**Comment by vanhauser-thc** [comment]
Is this currently waiting for the signal handling portion?
As this is a big patch cluster and it is already not merging because of the code changes I would find it preferable if this is merged sooner than later :)

And as our trunk is in "testing" anyway because of LTO this would be a perfect time to but afl-fuzz & co in testing as well.

**Comment by domenukk** [comment]
I've done my best to merge the changes.
As a quick&dirty fix for the custom mutator API drawbacks I posted in #191, I added an afl to each custom mutator call.
I'm pro merging this fast, keeping this branch up to date takes a long time

**Comment by vanhauser-thc** [comment]
go for it once you caught up. I will keep you busy today :p

**Comment by domenukk** [comment]
I've merged it. Fingers crossed :see_no_evil:

**Comment by vanhauser-thc** [comment]
it is colliding with a different issue - leak sanatizer that was not a problem is now reporting leaks everywhere and terminating travis early ... sigh

**Comment by domenukk** [comment]
In afl-fuzz three is a leak when argv @@ is replaced by the filename. I'm pretty sure this leak always existed... I've added a todo note there 😅
I didn't find the time to leak check afl-showmap, got to go to work now

**Comment by vanhauser-thc** [comment]
@domenukk afl-cmin fails (very likely because of afl-showmap)  and afl-tmin too - can you investigate?
you can see the failures in test.sh output

**Pull request merged**